datafusion/core/src/datasource/file_format/write/demux.rs

```diff
@@ -310,7 +311,37 @@ fn compute_partition_keys_by_row<'a>(
                 for i in 0..rb.num_rows() {
                     partition_values.push(array.value(i));
                 }
-            }
+            },
+            DataType::Dictionary(key_type, _) => {
+                match **key_type{
+                    DataType::UInt16 => {
```

Comment:
It would be nice to not have to enumerate every possible key_type here. It would be nice if there was a way to cast any Dictionary(_, V) -> V. Or at least iterate over any Dictionary(_, V) to pull out the V values in order without caring what type the key is.

Reply:
Hey Devin, there is a `downcast_dictionary_array!` macro that I encountered while attempting to fix this issue. (I was going to submit a PR but was figuring out if/how to add tests!) It lets you write code once that works for all key types, if I am not mistaken. I saw it used elsewhere in DataFusion code: https://github.com/apache/arrow-datafusion/blob/9fde5c4282fd9f0e3332fb40998bf1562c17fcda/datafusion/common/src/hash_utils.rs#L326-L329

You can see the code I used here: 
https://github.com/polygon-io/arrow-datafusion/blob/8c955891a6e5b6002faa07e54a67a6ec93347ade/datafusion/core/src/datasource/file_format/write/demux.rs#L316-L328
```rust
            DataType::Dictionary(_, value_type)
                if value_type.as_ref() == &DataType::Utf8 =>
            {
                downcast_dictionary_array!(
                    col_array =>  {
                        let array = col_array.downcast_dict::<StringArray>().unwrap();
                        for i in 0..rb.num_rows() {
                            partition_values.push(array.value(i));
                        }
                    },
                    _ => unreachable!(),
                )
            }
```

Reply:
Thanks @suremarc for the pointer to that macro! I updated this PR and it seems to work as expected!

I also added a new test based on a dictionary encoded parquet file from one of the testing submodules. It is working without error, but the values look a bit off. Not sure at this point if related to the downcasting or something else.